SRP test fails in Python 3.6 / Numpy 1.18+

The Streaming Random Patches test are failing for Python 3.6. This was confirmed in Linux and macOS:

```
=================================== FAILURES ===================================
_______ [doctest] river.ensemble.streaming_random_patches.SRPClassifier ________
075     >>> dataset = synth.ConceptDriftStream(seed=42, position=500,
076     ...                                    width=40).take(1000)
077     >>> model = ensemble.SRPClassifier(
078     ...     n_models=3,
079     ...     seed=42
080     ... )
081 
082     >>> metric = metrics.Accuracy()
083 
084     >>> evaluate.progressive_val_score(dataset, model, metric)
Expected:
    Accuracy: 76.78%
Got:
    Accuracy: 76.88%

```

--- a/river/ensemble/streaming_random_patches.py
+++ b/river/ensemble/streaming_random_patches.py
@@ -133,7 +133,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
         if warning_detector is None:
             warning_detector = ADWIN(delta=1e-4)
 
-        if disable_detector == 'off':
+        if disable_detector == "off":
             pass
         elif disable_detector == "drift":
             drift_detector = None
@@ -141,8 +141,10 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
         elif disable_detector == "warning":
             warning_detector = None
         else:
-            raise AttributeError(f"{disable_detector} is not a valid value for disable_detector.\n"
-                                 f"Valid options are: 'off', 'drift', 'warning'")
+            raise AttributeError(
+                f"{disable_detector} is not a valid value for disable_detector.\n"
+                f"Valid options are: 'off', 'drift', 'warning'"
+            )
         self.disable_detector = disable_detector
 
         if metric is None:
@@ -231,7 +233,9 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
             y_proba_temp = model.predict_proba_one(x)
             metric_value = model.metric.get()
             if not self.disable_weighted_vote and metric_value > 0.0:
-                y_proba_temp = {k: val * metric_value for k, val in y_proba_temp.items()}
+                y_proba_temp = {
+                    k: val * metric_value for k, val in y_proba_temp.items()
+                }
             y_pred.update(y_proba_temp)
 
         total = sum(y_pred.values())
@@ -312,7 +316,9 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
     def _init_ensemble(self, features: list):
         self._generate_subspaces(features=features)
 
-        subspace_indexes = np.arange(self.n_models)  # For matching subspaces with ensemble members
+        subspace_indexes = np.arange(
+            self.n_models
+        )  # For matching subspaces with ensemble members
         if (
             self.training_method == self._TRAIN_RANDOM_PATCHES
             or self.training_method == self._TRAIN_RANDOM_SUBSPACES
@@ -396,7 +402,9 @@ class StreamingRandomPatchesBaseLearner:
         self.n_warnings_detected = 0
 
         # Background learner
-        self._background_learner = None  # type: typing.Optional[StreamingRandomPatchesBaseLearner]
+        self._background_learner = (
+            None
+        )  # type: typing.Optional[StreamingRandomPatchesBaseLearner]
         self._background_learner_class = StreamingRandomPatchesBaseLearner
 
         # Nominal attributes
@@ -420,7 +428,9 @@ class StreamingRandomPatchesBaseLearner:
         if self.features is not None:
             # Select the subset of features to use
             x_subset = {k: x[k] for k in self.features}
-            if self._set_nominal_attributes and hasattr(self.model, "nominal_attributes"):
+            if self._set_nominal_attributes and hasattr(
+                self.model, "nominal_attributes"
+            ):
                 self.model.nominal_attributes = list(
                     set(self.features).intersection(set(self.nominal_attributes))
                 )
@@ -467,7 +477,9 @@ class StreamingRandomPatchesBaseLearner:
             if self.drift_detector.change_detected:
                 self.n_drifts_detected += 1
                 # There was a change, reset the model
-                self.reset(all_features=all_features, n_samples_seen=n_samples_seen, rng=rng)
+                self.reset(
+                    all_features=all_features, n_samples_seen=n_samples_seen, rng=rng
+                )
 
     def predict_proba_one(self, x):
         # Select the features to use
@@ -475,12 +487,16 @@ class StreamingRandomPatchesBaseLearner:
 
         return self.model.predict_proba_one(x_subset)
 
-    def _trigger_warning(self, all_features, n_samples_seen: int, rng: np.random.Generator):
+    def _trigger_warning(
+        self, all_features, n_samples_seen: int, rng: np.random.Generator
+    ):
         # Randomly generate a new subspace from all the original features
         subspace = (
             None
             if self.features is None
-            else random_subspace(all_features=all_features, k=len(self.features), rng=rng)
+            else random_subspace(
+                all_features=all_features, k=len(self.features), rng=rng
+            )
         )
 
         # Initialize the background learner
@@ -516,7 +532,9 @@ class StreamingRandomPatchesBaseLearner:
             subspace = (
                 None
                 if self.features is None
-                else random_subspace(all_features=all_features, k=len(self.features), rng=rng)
+                else random_subspace(
+                    all_features=all_features, k=len(self.features), rng=rng
+                )
             )
             # Reset model
             self.model = self.model.clone()

--- a/river/ensemble/streaming_random_patches.py
+++ b/river/ensemble/streaming_random_patches.py
@@ -423,7 +423,6 @@ class StreamingRandomPatchesBaseLearner:
         n_samples_seen: int,
         rng: np.random.Generator,
     ):
-        # print(sample_weight)
         all_features = [feature for feature in x.keys()]
         if self.features is not None:
             # Select the subset of features to use

--- a/river/ensemble/streaming_random_patches.py
+++ b/river/ensemble/streaming_random_patches.py
@@ -233,9 +233,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
             y_proba_temp = model.predict_proba_one(x)
             metric_value = model.metric.get()
             if not self.disable_weighted_vote and metric_value > 0.0:
-                y_proba_temp = {
-                    k: val * metric_value for k, val in y_proba_temp.items()
-                }
+                y_proba_temp = {k: val * metric_value for k, val in y_proba_temp.items()}
             y_pred.update(y_proba_temp)
 
         total = sum(y_pred.values())
@@ -316,9 +314,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
     def _init_ensemble(self, features: list):
         self._generate_subspaces(features=features)
 
-        subspace_indexes = np.arange(
-            self.n_models
-        )  # For matching subspaces with ensemble members
+        subspace_indexes = np.arange(self.n_models)  # For matching subspaces with ensemble members
         if (
             self.training_method == self._TRAIN_RANDOM_PATCHES
             or self.training_method == self._TRAIN_RANDOM_SUBSPACES
@@ -402,9 +398,7 @@ class StreamingRandomPatchesBaseLearner:
         self.n_warnings_detected = 0
 
         # Background learner
-        self._background_learner = (
-            None
-        )  # type: typing.Optional[StreamingRandomPatchesBaseLearner]
+        self._background_learner = None  # type: typing.Optional[StreamingRandomPatchesBaseLearner]
         self._background_learner_class = StreamingRandomPatchesBaseLearner
 
         # Nominal attributes
@@ -427,9 +421,7 @@ class StreamingRandomPatchesBaseLearner:
         if self.features is not None:
             # Select the subset of features to use
             x_subset = {k: x[k] for k in self.features}
-            if self._set_nominal_attributes and hasattr(
-                self.model, "nominal_attributes"
-            ):
+            if self._set_nominal_attributes and hasattr(self.model, "nominal_attributes"):
                 self.model.nominal_attributes = list(
                     set(self.features).intersection(set(self.nominal_attributes))
                 )
@@ -476,9 +468,7 @@ class StreamingRandomPatchesBaseLearner:
             if self.drift_detector.change_detected:
                 self.n_drifts_detected += 1
                 # There was a change, reset the model
-                self.reset(
-                    all_features=all_features, n_samples_seen=n_samples_seen, rng=rng
-                )
+                self.reset(all_features=all_features, n_samples_seen=n_samples_seen, rng=rng)
 
     def predict_proba_one(self, x):
         # Select the features to use
@@ -486,16 +476,12 @@ class StreamingRandomPatchesBaseLearner:
 
         return self.model.predict_proba_one(x_subset)
 
-    def _trigger_warning(
-        self, all_features, n_samples_seen: int, rng: np.random.Generator
-    ):
+    def _trigger_warning(self, all_features, n_samples_seen: int, rng: np.random.Generator):
         # Randomly generate a new subspace from all the original features
         subspace = (
             None
             if self.features is None
-            else random_subspace(
-                all_features=all_features, k=len(self.features), rng=rng
-            )
+            else random_subspace(all_features=all_features, k=len(self.features), rng=rng)
         )
 
         # Initialize the background learner
@@ -531,9 +517,7 @@ class StreamingRandomPatchesBaseLearner:
             subspace = (
                 None
                 if self.features is None
-                else random_subspace(
-                    all_features=all_features, k=len(self.features), rng=rng
-                )
+                else random_subspace(all_features=all_features, k=len(self.features), rng=rng)
             )
             # Reset model
             self.model = self.model.clone()

--- a/river/ensemble/streaming_random_patches.py
+++ b/river/ensemble/streaming_random_patches.py
@@ -46,11 +46,15 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
     lam
         Lambda value for resampling.
     drift_detector
-        Drift detector. If `None`, disables concept drift detection and the
-        background learner.
+        Drift detector.
     warning_detector
-        Warning detector. If `None`, disables the background learner and
-        ensemble members are reset if drift is detected.
+        Warning detector.
+    disable_detector:
+        Option to disable drift detectors:<br/>
+        * If `'off'`, detectors are enabled.<br/>
+        * If `'drift'`, disables concept drift detection and the background learner.<br/>
+        * If `'warning'`, disables the background learner and ensemble members are
+         reset if drift is detected.<br/>
     disable_weighted_vote
         If True, disables weighted voting.
     nominal_attributes
@@ -114,6 +118,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
         lam: float = 6.0,
         drift_detector: base.DriftDetector = None,
         warning_detector: base.DriftDetector = None,
+        disable_detector: str = "off",
         disable_weighted_vote: bool = False,
         nominal_attributes=None,
         seed=None,
@@ -128,6 +133,17 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
 
         if warning_detector is None:
             warning_detector = ADWIN(delta=1e-4)
+
+        if disable_detector == 'off':
+            pass
+        elif disable_detector == "drift":
+            drift_detector = None
+            warning_detector = None
+        elif disable_detector == "warning":
+            warning_detector = None
+        else:
+            raise AttributeError(f"{disable_detector} is not a valid value for disable_detector.\n"
+                                 f"Valid options are: 'off', 'drift', 'warning'")
 
         if metric is None:
             metric = Accuracy()
@@ -184,7 +200,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
             # i.e. all instances are used for training.
             if self.training_method == self._TRAIN_RANDOM_SUBSPACES:
                 model.learn_one(
-                    X=x,
+                    x=x,
                     y=y,
                     sample_weight=1.0,
                     n_samples_seen=self._n_samples_seen,

--- a/river/ensemble/streaming_random_patches.py
+++ b/river/ensemble/streaming_random_patches.py
@@ -11,7 +11,6 @@ from river.drift import ADWIN
 from river.metrics import Accuracy
 from river.metrics.base import MultiClassMetric
 from river.tree import HoeffdingTreeClassifier
-from river.utils.skmultiflow_utils import check_random_state
 
 
 class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
@@ -62,10 +61,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
         attributes are numerical. Note: Only applies if the base model
         allows to define the nominal attributes.
     seed
-        If `int`, `seed` is used to seed the random number generator;
-        If `RandomState`, `seed` is the random number generator;
-        If `None`, the random number generator is the `RandomState` instance
-        used by `np.random`.
+        Random number generator seed for reproducibility.
     metric
         Metric to track members performance within the ensemble.
 
@@ -73,20 +69,23 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
     --------
     >>> from river import synth
     >>> from river import ensemble
+    >>> from river import tree
     >>> from river import evaluate
     >>> from river import metrics
 
     >>> dataset = synth.ConceptDriftStream(seed=42, position=500,
-    ...                                    width=40).take(1000)
-    >>> model = ensemble.SRPClassifier(
-    ...     n_models=3,
-    ...     seed=42
+    ...                                    width=50).take(1000)
+    >>> base_model = tree.HoeffdingTreeClassifier(
+    ...     grace_period=50, split_confidence=0.01,
+    ...     nominal_attributes=['age', 'car', 'zipcode']
     ... )
-
+    >>> model = ensemble.SRPClassifier(
+    ...     model=base_model, n_models=3, seed=42,
+    ... )
     >>> metric = metrics.Accuracy()
 
     >>> evaluate.progressive_val_score(dataset, model, metric)
-    Accuracy: 76.78%
+    Accuracy: 71.17%
 
     References
     ----------
@@ -144,6 +143,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
         else:
             raise AttributeError(f"{disable_detector} is not a valid value for disable_detector.\n"
                                  f"Valid options are: 'off', 'drift', 'warning'")
+        self.disable_detector = disable_detector
 
         if metric is None:
             metric = Accuracy()
@@ -161,7 +161,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
         self.metric = metric
         self.nominal_attributes = nominal_attributes if nominal_attributes else []
         self.seed = seed
-        self._rng = check_random_state(self.seed)
+        self._rng = np.random.default_rng(self.seed)
 
         self._n_samples_seen = 0
         self._subspaces = None
@@ -202,7 +202,7 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
                 model.learn_one(
                     x=x,
                     y=y,
-                    sample_weight=1.0,
+                    sample_weight=1,
                     n_samples_seen=self._n_samples_seen,
                     rng=self._rng,
                 )
@@ -333,16 +333,16 @@ class SRPClassifier(base.WrapperMixin, base.EnsembleMixin, base.Classifier):
                     drift_detector=self.drift_detector,
                     warning_detector=self.warning_detector,
                     is_background_learner=False,
+                    rng=self._rng,
                     features=subspace,
                     nominal_attributes=self.nominal_attributes,
-                    rng=self._rng,
                 )
             )
 
     def reset(self):
         self.models = []
         self._n_samples_seen = 0
-        self._rng = check_random_state(self.seed)
+        self._rng = np.random.default_rng(self.seed)
 
 
 class StreamingRandomPatchesBaseLearner:
@@ -359,9 +359,9 @@ class StreamingRandomPatchesBaseLearner:
         drift_detector: base.DriftDetector,
         warning_detector: base.DriftDetector,
         is_background_learner,
+        rng: np.random.Generator,
         features=None,
         nominal_attributes=None,
-        rng=None,
     ):
         self.idx_original = idx_original
         self.created_on = created_on
@@ -393,9 +393,7 @@ class StreamingRandomPatchesBaseLearner:
 
         # Statistics
         self.n_drifts_detected = 0
-        self.n_drifts_induced = 0
         self.n_warnings_detected = 0
-        self.n_warnings_induced = 0
 
         # Background learner
         self._background_learner = None  # type: typing.Optional[StreamingRandomPatchesBaseLearner]
@@ -406,7 +404,7 @@ class StreamingRandomPatchesBaseLearner:
         self._set_nominal_attributes = self._can_set_nominal_attributes()
 
         # Random number generator (initialized)
-        self._rng = rng
+        self.rng = rng
 
     def learn_one(
         self,
@@ -415,8 +413,9 @@ class StreamingRandomPatchesBaseLearner:
         *,
         sample_weight: int,
         n_samples_seen: int,
-        rng: np.random.RandomState,
+        rng: np.random.Generator,
     ):
+        # print(sample_weight)
         all_features = [feature for feature in x.keys()]
         if self.features is not None:
             # Select the subset of features to use
@@ -476,7 +475,7 @@ class StreamingRandomPatchesBaseLearner:
 
         return self.model.predict_proba_one(x_subset)
 
-    def _trigger_warning(self, all_features, n_samples_seen: int, rng: np.random.RandomState):
+    def _trigger_warning(self, all_features, n_samples_seen: int, rng: np.random.Generator):
         # Randomly generate a new subspace from all the original features
         subspace = (
             None
@@ -493,20 +492,14 @@ class StreamingRandomPatchesBaseLearner:
             drift_detector=self.drift_detector,
             warning_detector=self.warning_detector,
             is_background_learner=True,
+            rng=self.rng,
             features=subspace,
             nominal_attributes=self.nominal_attributes,
-            rng=self._rng,
         )
         # Hard-reset the warning method
         self.warning_detector = self.warning_detector.clone()
 
-    def reset(self, all_features: list, n_samples_seen: int, rng: np.random.RandomState):
-        # Randomly generate a new subspace from all the original features
-        subspace = (
-            None
-            if self.features is None
-            else random_subspace(all_features=all_features, k=len(self.features), rng=rng)
-        )
+    def reset(self, all_features: list, n_samples_seen: int, rng: np.random.Generator):
 
         if not self.disable_background_learner and self._background_learner is not None:
             # Replace model with the corresponding background model
@@ -519,6 +512,12 @@ class StreamingRandomPatchesBaseLearner:
             self.features = self._background_learner.features
             self._background_learner = None
         else:
+            # Randomly generate a new subspace from all the original features
+            subspace = (
+                None
+                if self.features is None
+                else random_subspace(all_features=all_features, k=len(self.features), rng=rng)
+            )
             # Reset model
             self.model = self.model.clone()
             self.metric.cm.reset()
@@ -531,7 +530,7 @@ class StreamingRandomPatchesBaseLearner:
         return self.nominal_attributes is not None and len(self.nominal_attributes) > 0
 
 
-def random_subspace(all_features: list, k: int, rng: np.random.RandomState):
+def random_subspace(all_features: list, k: int, rng: np.random.Generator):
     """Utility function to generate a random feature subspace of length k
 
     Parameters

